### ISSUE TITLE
Allow larger / custom render distances (for eg. DH)

### ISSUE THREAD

**Issue by MysticMia** [Qendolin/better-clouds#76]
**Is your feature request related to a problem? Please describe.**
Currently, the maximum distance of clouds is limited at 4x. Since Distant Horizons has a custom render distance method, it renders blocks much farther out than 4x your vanilla render distance. When flying through the sky using a mod like Distant Horizons, you can notice the clouds rendering in a circle around the player, but the world extending quite a bit beyond the clouds. This is more obvious when moving around.

![image](https://github.com/Qendolin/better-clouds/assets/65396931/66b304e4-d70b-4d92-a649-8222237d4d44)

> Notice the smaller render distance of the clouds. You can see distant buildings going above where clouds would have been.

![image](https://github.com/Qendolin/better-clouds/assets/65396931/f2e910c0-6655-48c4-8c91-da6ee85109fa)

> Picture at y=500, where the range of chunks exceeds the range of the clouds. 

**Describe the solution you'd like**
Set a higher limit for the render distance multiplier, allowing players to raise it to >4 (10x? 40x? a textbox input with custom limit?)

**Additional context**
Another mod that extends render distances, Bobby, uses the vanilla slider to extend render distances. Using this mod, you can set the render distances up to 1024 chunks, for example. Using this mod, the clouds still render at 4x distance (going up to 4096 chunks, I assume) (although it does take ~5 seconds to load the clouds, but no lag)
![image](https://github.com/Qendolin/better-clouds/assets/65396931/841fa56c-5c39-4461-89ca-5127eafc4b7c)

> Notice the large render distance for the clouds

**Describe alternatives you've considered**
Since Distant Horizons has a secondary render distance slider, it may be better not to render the distant clouds at the same resolution: render them at lower quality.

I looked at older issues and found #45. I installed the Extended Clouds mod and compared the two cloud render distances:
![image](https://github.com/Qendolin/better-clouds/assets/65396931/47739ba7-e918-4bb3-83e0-50b4a2fba03b)

> Clouds while selecting "Better Clouds" in the mod config

![image](https://github.com/Qendolin/better-clouds/assets/65396931/013c70ae-3c97-4d01-8a4f-978fa24ff0e8)

> Clouds when using "Vanilla" cloud rendering. It seems to be independent from the 4x render distance limit, going up to a distance seemingly 600 to 800 chunks away from the player (10000 to 14000 blocks? hard to estimate).

This was with config `"cloudRenderDistanceMultiplier": 1.0,` in extended-clouds.json. The clouds disappeared going above a height of 4892 (with a cloud height of 240) (cloud height-dependent).

**Comment by Qendolin** [comment]
I'll keep this issue open, but I don't think it'll ever get done. It's too much work.
Here are some of the problems that I see:
## It doesn't look good
I don't like that it gets very "dense" at the edge.
![Far clouds](https://private-user-images.githubusercontent.com/65396931/307877255-841fa56c-5c39-4461-89ca-5127eafc4b7c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5NzYzNDEsIm5iZiI6MTcwODk3NjA0MSwicGF0aCI6Ii82NTM5NjkzMS8zMDc4NzcyNTUtODQxZmE1NmMtNWMzOS00NDYxLTg5Y2EtNTEyN2VhZmM0YjdjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDE5MzQwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3ODE4M2ViOTRjNWJhZDEwOGIzMTZiMGJmNGIyMzY4N2Q5ODU2YzNiYTA3NGQzOTM3ZmZlNzQyY2MwYzRlN2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ecClC818ZJfUSboeXlygWpS7zxZ0vWs4iTaffgU9JyE)

Also when viewed from above there is not enough variation at a large scale.

## Generation performance
Like you said, it takes a while to generate all those points. This would require significant changes in the generator.

## Render performance
It's hard to say how much the impact would be. But rendering a lot, and it really is a lot, of anything that has a lot of overlap is not liked by the GPU. For far away clouds I'ld have to come up with some way to prevent the more-than-linear increase in "number of clouds" to "render time"

Maybe I could just render a single large, flat texture at a certain distance. If the clouds are far enough away their height is not really perceivable anyway.

## Code complexity
Adding optional code for rendering at a larger distance makes everything more messy.

**Comment by obj-obj** [comment]
I mean, you can always simply increase the limit for the current render distance without any other changes

**Comment by Qendolin** [comment]
And how would that work without running into any of the problems stated above?

**Comment by obj-obj** [comment]
> And how would that work without running into any of the problems stated above?

The screenshots sent by @MysticMia look perfectly fine to me. Just say a view distance of more than 4x the render distance is unsupported

**Comment by Qendolin** [comment]
It is unsupported because it doesn't just work. The screenshot might look fine but believe me, simply increasing the distance causes visual bugs as well as too bad generation performance. Beyond the limit the GPU can't differentiate between near and far objects, leading to stuff being rendered in the wrong order.

**Comment by offbem** [comment]
@Qendolin  Couldn't you contact the developers of Distant Horizon and have clouds generate in LODS in addition to the default rendering?

**Comment by Qendolin** [comment]
No, DH does not manage the generation of this mod's clouds. There is nothing that DH needs to do.

**Comment by Foozey** [comment]
I would love to see the option for farther cloud distance. It often looks unusually short if using DH. Perhaps you could detach the cloud distance from render distance and instead allow the user to set a value within a specific range that you're comfortable with allowing?

For example, I play on 16 render distance with 256 DH, but even with the max 4x in better clouds, the distance looks a bit short. If I up the render distance to 32, the cloud distance looks much more appealing. This is why I think detaching it could be a good idea, so someone could use a longer cloud distance but still play on lower render distances.

**Comment by Qendolin** [comment]
As I've said before, it's not so simple. 1.21.6 improved things, because there cloud distance is decoupled from view distance by the game itself. But going even further still isn't possible with how the mod is currently programmed. And I don't currently feel any motivation to add new features to the mod. I have a very experimental branch https://github.com/Qendolin/better-clouds/tree/far-rendering that rewrites the rendering, to allow for further distances, but it's far from done.

**Comment by orangishcat** [comment]
Unfortunately it's not worth the huge increase in complexity. The only other viable option is to brute-force increase the vanilla cloud distance option, which should work but runs into transparency and performance issues. I don't think this will ever be fixed.
